static/app/utils/profiling/renderers/flamegraphRenderer.tsx

```diff
@@ -0,0 +1,675 @@
+  makeProjectionMatrix,
+  Rect,
+  resizeCanvasToDisplaySize,
+} from '../gl/utils';
+
+import {fragment, vertex} from './shaders';
+
+export function uniqueFrameKey(frame: FlamegraphFrame): string {
+  return `${frame.frame.key + String(frame.start)}`;
+}
+
+class FlamegraphRenderer {
+  canvas: HTMLCanvasElement | null;
+  flamegraph: Flamegraph;
+
+  gl: WebGLRenderingContext | null = null;
+  program: WebGLProgram | null = null;
+
+  theme: FlamegraphTheme;
+  origin: vec2;
+  frames: ReadonlyArray<FlamegraphFrame> = [];
+  roots: ReadonlyArray<FlamegraphFrame> = [];
+
+  // Vertex and color buffer
+  positions: Array<number> = [];
+  bounds: Array<number> = [];
+  colors: Array<number> = [];
+
+  colorMap: Map<string | number, number[]> = new Map();
+
+  logicalSpace: Rect = Rect.Empty();
+  logicalToPhysicalSpace: mat3 = mat3.identity(mat3.create());
+
+  physicalSpace: Rect = new Rect(0, 0, 0, 0);
+  physicalToLogicalSpace: mat3 = mat3.identity(mat3.create());
+
+  configSpace: Rect = new Rect(0, 0, 0, 0);
+  configView: Rect = new Rect(0, 0, 0, 0);
+
+  lastDragPosition: vec2 | null = null;
+
+  attributes: {
+    a_position: number | null;
+    a_color: number | null;
+    a_bounds: number | null;
+  } = {
+    a_position: null,
+    a_color: null,
+    a_bounds: null,
+  };
+
+  uniforms: {
+    u_projection: WebGLUniformLocation | null;
+    u_model: WebGLUniformLocation | null;
+    u_is_search_result: WebGLUniformLocation | null;
+    u_border_width: WebGLUniformLocation | null;
+    u_draw_border: WebGLUniformLocation | null;
+  } = {
+    u_border_width: null,
+    u_draw_border: null,
+    u_is_search_result: null,
+    u_model: null,
+    u_projection: null,
+  };
+
+  options: {
+    draw_border: boolean;
+  };
+
+  constructor(
+    canvas: HTMLCanvasElement,
+    flamegraph: Flamegraph,
+    theme: FlamegraphTheme,
+    origin: vec2 = vec2.fromValues(0, 0),
+    options: {draw_border: boolean} = {draw_border: false}
+  ) {
+    this.flamegraph = flamegraph;
+    this.canvas = canvas;
+    this.theme = theme;
+    this.origin = origin;
+    this.options = options;
+
+    this.init();
+  }
+
+  init(): void {
+    const VERTICES = 6;
+    const COLOR_COMPONENTS = 4;
+
+    this.colors = new Array(VERTICES * COLOR_COMPONENTS);
+    this.frames = [...this.flamegraph.frames];
+    this.roots = this.flamegraph.frames.filter(f => !f.parent);
+
+    // If our flamegraph has color encoding, then just pull the colors into the color buffer.
+    // This happens with differential flamegraphs, where the color changes are encoded at the time
+    // when we construct the differential flamegraph.
+    if (this.flamegraph.colors) {
+      this.colorMap = this.flamegraph.colors;
+
+      const length = this.frames.length;
+
+      for (let index = 0; index < length; index++) {
+        const frame = this.frames[index];
+        const c = this.colorMap.get(
+          frame.frame.name + (frame.frame.file ? frame.frame.file : '')
```

Comment:
@Zylphrex 100%, lets just consolidate this a bit with the function that generates colors for the flamegraph, because they need to match. So the key is sort of dynamic, which is why I just left it like this bc I wasnt sure what the best course would be